### PR TITLE
toXXAscii use xor op, saving 30%~50% time

### PR DESCRIPTION
**Pull request by bung87**
outputs:  
```
CPU time xor 0.012819
CPU time toLowerAscii 0.040421
xor ver time consuming is about 31.71371316889735% of toLowerAscii
CPU time string strtoLowerAscii 0.151566
CPU time string strtoUpperAscii 0.149303
CPU time string strxorLower 0.07348199999999999
CPU time string strxorUpper 0.07298599999999994
```
code validation below  

``` nim


block toupper:
  var h: char
  for c in 'a' .. 'z':
    h = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
    doAssert h == c.toUpperAscii

block tolower:
  var h: char
  for c in 'A' .. 'Z':
    h = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
    doAssert h == c.toLowerAscii

var t0 = cpuTime()
var d: char
for i in 1 .. 100_000:
  for c in 'A' .. 'Z':
    d = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
var r1 = cpuTime() - t0
echo "CPU time xor ", r1

var t1 = cpuTime()
var e: char
for i in 1 .. 100_000:
  for c in 'A' .. 'Z':
    e = toLowerAscii(c)
var r2 = cpuTime() - t1
echo "CPU time toLowerAscii ", r2
doAssert r1 < r2
echo "xor ver time consuming is about " & $( (r1 / r2) * 100) & "%" & " of toLowerAscii"

const All = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"#$%&\'()*+,-./:;?@[\\]^_`{|}~ \t\n\r\x0b\x0c"

block strtoLowerAscii:
  var t3 = cpuTime()
  var e: char
  for i in 1 .. 100_000:
    for c in All:
      e = toLowerAscii(c)
  var t4 = cpuTime() - t3
  echo "CPU time string strtoLowerAscii ", t4

block strtoUpperAscii:
  var t3 = cpuTime()
  var e: char
  for i in 1 .. 100_000:
    for c in All:
      e = toUpperAscii(c)
  var t4 = cpuTime() - t3
  echo "CPU time string strtoUpperAscii ", t4

block strxorLower:
  var t3 = cpuTime()
  var e: char
  for i in 1 .. 100_000:
    for c in All:
      if c in {'A' .. 'Z'}:
      # if c >= 'A' and c <= 'Z' :
        e = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
      else:
        e = c
  var t4 = cpuTime() - t3
  echo "CPU time string strxorLower ", t4

  block strxorUpper:
    var t3 = cpuTime()
    var e: char
    for i in 1 .. 100_000:
      for c in All:
        if c in {'a' .. 'z'}:
        # if c >= 'a' and c <= 'z':
          e = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
        else:
          e = c
    var t4 = cpuTime() - t3
    echo "CPU time string strxorUpper ", t4

```

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -209,7 +209,7 @@ proc toLowerAscii*(c: char): char {.noSideEffect,
     doAssert toLowerAscii('A') == 'a'
     doAssert toLowerAscii('e') == 'e'
   if c in {'A'..'Z'}:
-    result = chr(ord(c) + (ord('a') - ord('A')))
+    result = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
```
Comment (metagn) [inline lib/pure/strutils.nim]:
```suggestion
    result = char(uint8(c) xor 0b0010_0000'u8)
```

Comment (bung87) [reply lib/pure/strutils.nim]: why do convert ? casting is not just fine there?

Comment (metagn) [reply lib/pure/strutils.nim]: It just looks ugly, and should really only be used for unsafe memory operations afaik. Char and uint8 are the same type in representation so it doesn't functionally change anything here

Comment (bung87) [reply lib/pure/strutils.nim]: I prefer cast, when you do convert in other types they do different operation produce different results. cast match my intention, I dont meant to convert one data type to other data type.

File: lib/pure/strutils.nim [hunk 2]
```diff
@@ -248,7 +248,7 @@ proc toUpperAscii*(c: char): char {.noSideEffect,
     doAssert toUpperAscii('a') == 'A'
     doAssert toUpperAscii('E') == 'E'
   if c in {'a'..'z'}:
-    result = chr(ord(c) - (ord('a') - ord('A')))
+    result = cast[char](cast[uint8](c) xor 0b0010_0000'u8)
```
Comment (Araq) [inline lib/pure/strutils.nim]:
```suggestion
    result = char(uint8(c) xor 0b0010_0000'u8)
```

**Pull request merged**